internal/pkg/propagation/summaries.go

```diff
@@ -0,0 +1,761 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package propagation
+
+// A summary captures the behavior of a function with respect to taint
+// propagation. Specifically, given that at least one of the necessary
+// arguments is tainted, which arguments/return values become tainted?
+// Note that when it's present, the receiver counts as an argument.
+//
+// As an example, consider fmt.Fprintf:
+//   func Fprintf(w io.Writer, format string, a ...interface{}) (n int, err error) {
+// Its summary is:
+//   "fmt.Fprintf": {
+//   	ifTainted:   0b110,
+//   	taintedArgs: []int{0},
+//   },
+// In English, this says that if the format string or the varargs slice are
+// tainted, then the Writer is tainted.
+type summary struct {
+	// ifTainted is a bitset which contains positions for parameters
+	// such that if one of these parameters is tainted, taint should
+	// be propagated to the arguments and return values.
+	// There is a 1-to-1 mapping between the bits and the function's
+	// parameters, with the least significant bit corresponding to the
+	// first (0th) argument.
+	ifTainted int64
```

Comment:
This feels like a premature optimization.  I don't see why this is a bitwise `int64` but `taintedArgs` / `taintedRets` are `[]int`.

If you're worried about performance, a `map[int]struct{args, rets []int}` mapping input tainted-arg to output tainted-whatever would still be well usable and reasonably readable.  I don't see this struct being a performance or memory bottleneck, though.

------

Coming back to this now that I've read everything else, it's not that bad.  I think it's more the mismatch that's bothering me.  I like the utility of `fromFirstArgToFirstRet`, and I feel like that could be generalized into something like

```golang
const (
  first int64 = 1 << 1
  second
  third
  fourth
)

var funcSummaries = map[string]summary{
  // func Errorf(format string, a ...interface{}) error {
  "fmt.Errorf": {
    ifTainted:   first | second,
    taintedRets: first,
  },
  // ...
}
```

Either way, I think I'd like there to be consistency between the "input slice" and "output slices".

Reply:
(I don't anticipate performance to be an issue here. FWIW, and I should probably have mentioned this somewhere, AFAICT these changes have no performance impact.) I typically keep bitwise manipulation as a last resort, but in this case I think it leads to the most readable code, specifically because it makes it very easy to test whether the intersection is empty or not.

I agree that it would be nice to have consistency between the inputs and outputs, but I think it makes more sense to have separate data representations because they are used in different ways. The bitset makes sense when looking for taint because we're interested in the intersection. The slices make sense when looking for where to propagate taint, because we can just iterate over them.

I really like your `first, second, third, ...` idea for the bits. I think this is a great improvement to readability. I have implemented it.

Reply:
Glad you adopted the `first | second` approach.  I still think it would read a bit better if the other fields were consistent with it, but I'm not going to bikeshed on it.  Marking this thread resolved.